feat(semantic): add ability to lookup if AST contains any node kinds

camchenry opened the pull request:
I decided to revive this stack: https://github.com/oxc-project/oxc/pull/12227 and merge all of the changes into one PR for slightly easier review and changes.

This PR adds a new `AstTypesBitset` struct which is a set of bits that has a fixed size known at compile time and can be used to store AST types as a set. The aim of this structure is to making checks like "does this file have a for statement?" possible to answer without doing any additional iterations over the AST.

camchenry commented:
* **#13137** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13137?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13137?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13137).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_semantic/src/ast_types_bitset.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// Returns `true` if all bits in `other` are set in `self`.
+    pub fn contains(&self, other: &Self) -> bool {
+        for (&a, &b) in self.0.iter().zip(other.0.iter()) {
+            if a & b != b {
+                return false;
+            }
+        }
+        true
+    }
```
Probably preferable to make this branchless:

```rs
pub fn contains(&self, other: &Self) -> bool {
    let mut mismatches = 0;
    for (&a, &b) in self.0.iter().zip(other.0.iter()) {
        let set_in_both = a & b;
        // 0 if `set_in_both == b`
        let mismatch = set_in_both ^ b;
        mismatches |= mismatch;
    }
    mismatches == 0
}
```

https://godbolt.org/z/19sv53KMM

overlookmotel commented on crates/oxc_semantic/src/node.rs:
```diff
@@ -242,6 +250,33 @@ impl<'a> AstNodes<'a> {
+
+    /// Checks if the AST contains any nodes of the given types.
+    ///
+    /// Example:
+    /// - `.contains_any(&[AstType::ForStatement])` returns `true` if there is a `for` loop anywhere in the AST.
+    /// - `.contains_any(&[AstType::ImportDeclaration, AstType::ExportDeclaration])` returns `true` if there are any imports OR exports in the AST.
```
These examples are out of date. Ditto `contains_all`.

overlookmotel commented on crates/oxc_semantic/src/node.rs:
```diff
@@ -234,6 +241,7 @@ impl<'a> AstNodes<'a> {
         );
         self.parent_ids.push(NodeId::ROOT);
         self.nodes.push(AstNode::new(kind, scope_id, cfg_id, flags, NodeId::ROOT));
+        self.node_kinds_set.set(kind.ty());
```
```suggestion
        self.node_kinds_set.set(AstKind::Program);
```

Or, since every AST contains `Program`, could just initialize `AstTypesBitset` at the start with that bit already set.

camchenry replied:
> Or, since every AST contains `Program`, could just initialize `AstTypesBitset` at the start with that bit already set.

Yeah this did occur to me. Maybe we'll do it in a follow-up PR, it's probably very minor at the moment.

camc314 reviewed (commented):
this looks really nice, should definetly help optimizing linter perf

overlookmotel reviewed (approved):
Feel free to correct the comment in a later PR, rather than having to restack this whole stack.

overlookmotel commented on crates/oxc_semantic/src/ast_types_bitset.rs:
```diff
@@ -0,0 +1,116 @@
+
+const USIZE_BITS: usize = usize::BITS as usize;
+
+/// Number of bytes required for bit set which can represent all [`AstType`]s.
```
I'll submit a follow-on PR for this change.

overlookmotel replied:
#13572

camchenry committed:
feat(semantic): add ability to lookup if AST contains any node kinds (#13137)

I decided to revive this stack: https://github.com/oxc-project/oxc/pull/12227 and merge all of the changes into one PR for slightly easier review and changes.

This PR adds a new `AstTypesBitset` struct which is a set of bits that has a fixed size known at compile time and can be used to store AST types as a set. The aim of this structure is to making checks like "does this file have a for statement?" possible to answer without doing any additional iterations over the AST.
